Markup case names

flooie opened the pull request:
@grossir 

This PR adds support for identifying the full case name in markup text.  When a plaintiff is identified
it calls a new method to check if the text is inside a tag and extracts that out. 

Also supports antecedent guess

This should improve reference citation extraction and better identify citations.

grossir commented on eyecite/helpers.py:
```diff
@@ -129,13 +132,64 @@ def add_post_citation(citation: CaseCitation, words: Tokens) -> None:
 
 
-def add_defendant(citation: CaseCitation, words: Tokens) -> None:
+def update_text_from_markup(
```
the name should be more specific?  something like  `update_citation_metadata_from_markup`?

flooie replied:
sure thanks

grossir commented on eyecite/helpers.py:
```diff
@@ -129,13 +132,64 @@ def add_post_citation(citation: CaseCitation, words: Tokens) -> None:
+        if tag_start <= markup_start and markup_end <= tag_end:
+            start = document.markup_to_plain.update(tag_start, bisect_right)
+            text = document.plain_text[start:end].strip()
+            if short:
+                citation.metadata.antecedent_guess = text
+            else:
+                citation.metadata.plaintiff = text
```
You also need to update the full span start and end, to resolve overlaps

grossir commented on eyecite/helpers.py:
```diff
@@ -129,13 +132,64 @@ def add_post_citation(citation: CaseCitation, words: Tokens) -> None:
+
+    doc = html.fromstring(document.markup_text)
+
+    for elem in doc.xpath("//em | //i"):
```
I think this may be slow for bigger documents. 
You are just interested in the inmediate vicinity of the name next to the citation, so why check every single tag every time?

flooie replied:
or maybe once at the start of the extraction?

grossir reviewed (commented):
looking good, some comments below

flooie commented:
```
---

Co.                                                                    --->      Silverthorne Lumber Co.
Silverthorne Lumber Co. v. United States, 251 U. S. 385
---

Portelli, 15 N. Y. 2d 235                                              --->      Portelli
People v. Portelli, 15 N. Y. 2d 235, 205 N. E. 2d 857
---

States                                                                 --->      United States
United States v. Grunewald, 233 F. 2d 556
---

States                                                                 --->      United States
Grunewald, 233 F. 2d 556, 579, 581-582 (Frank, J., dissenting), rev'd  --->      Grunewald
United States v. Grunewald, 233 F. 2d 556, 579, 581-582 (Frank, J., dissenting), rev'd, 353 U. S. 391
---

Dorado, 62 Cal. 2d 338, 351                                            --->      Dorado
People v. Dorado, 62 Cal. 2d 338, 351, 398 P. 2d 361
---

Dorado, 62 Cal. 2d 338, 351, 398 P. 2d 361, 369-370                    --->      Dorado
People v. Dorado, 62 Cal. 2d 338, 351, 398 P. 2d 361, 369-370, 42 Cal. Rptr. 169
---

U. S., 337 F. 2d 136 (1964), cert. den.                                --->      U. S.
Jackson v. U. S., 337 F. 2d 136 (1964), cert. den. 380 U. S. 935
---

S.                                                                     --->      U. S.
U. S. v. Konigsberg, 336 F. 2d 844
---

S.                                                                     --->      U. S.
Konigsberg, 336 F. 2d 844 (1964), cert. den.                           --->      Konigsberg
U. S. v. Konigsberg, 336 F. 2d 844 (1964), cert. den. 379 U. S. 933
---

States                                                                 --->      United States
United States v. Scully, 225 F. 2d 113
---

States                                                                 --->      United States
Carignan, *528                                                         --->      Carignan
United States v. Carignan, *528 342 U. S. 36
---

Hansen                                                                 --->      Sparf and Hansen
Sparf and Hansen v. United States, 156 U. S. 51
---

States                                                                 --->      United States
United States v. Mitchell, 322 U. S. 65
---

States                                                                 --->      United States
United States v. Carignan, 342 U. S. 36
---

States                                                                 --->      United States
United States v. Romano, 382 U. S. 136
---

State, 238 Md. 140                                                     --->      State
Johnson v. State, 238 Md. 140, 207 A. 2d 643
---

Modesto, 62 Cal. 2d 436, 446                                           --->      Modesto
People v. Modesto, 62 Cal. 2d 436, 446, 398 P. 2d 753
---

States                                                                 --->      United States
United States v. Drummond, 354 F. 2d 132
---

States                                                                 --->      United States
United States v. Childress, 347 F. 2d 448
---

Dorado, 62 Cal. 2d 338                                                 --->      Dorado
People v. Dorado, 62 Cal. 2d 338, 398 P. 2d 361
---

Dorado, 62 Cal. 2d 338, 398 P. 2d 361                                  --->      Dorado
People v. Dorado, 62 Cal. 2d 338, 398 P. 2d 361, 42 Cal. Rptr. 169
---

Hartgraves, 31 Ill. 2d 375                                             --->      Hartgraves
People v. Hartgraves, 31 Ill. 2d 375, 202 N. E. 2d 33
---

Wakat, 415 Ill. 610                                                    --->      Wakat
People v. Wakat, 415 Ill. 610, 114 N. E. 2d 706
---

State, 213 Md. 556                                                     --->      State
Kier v. State, 213 Md. 556, 132 A. 2d 494
---

People, 113 Colo. 194                                                  --->      People
Bruner v. People, 113 Colo. 194, 156 P. 2d 111
---

Matlock, 51 Cal. 2d 682                                                --->      Matlock
People v. Matlock, 51 Cal. 2d 682, 336 P. 2d 505
---

States                                                                 --->      United States
United States v. Murphy, 222 F. 2d 698
---

Bonino, 1 N. Y. 2d 752                                                 --->      Bonino
People v. Bonino, 1 N. Y. 2d 752, 135 N. E. 2d 51
---

States                                                                 --->      United States
United States v. Carignan, 342 U. S. 36
---

Donovan, 13 N. Y. 2d 148                                               --->      Donovan
People v. Donovan, 13 N. Y. 2d 148, 193 N. E. 2d 628
---

Brown                                                                  --->      United States ex rel. Brown
United States ex rel. Brown v. Fay, 242 F. Supp. 273
---

Witenski, 15 N. Y. 2d 392                                              --->      Witenski
People v. Witenski, 15 N. Y. 2d 392, 207 N. E. 2d 358
---

Dorado, 62 Cal. 2d 338, 354                                            --->      Dorado
People v. Dorado, 62 Cal. 2d 338, 354, 398 P. 2d 361
---

Dorado, 62 Cal. 2d 338, 354, 398 P. 2d 361, 371                        --->      Dorado
People v. Dorado, 62 Cal. 2d 338, 354, 398 P. 2d 361, 371, 42 Cal. Rptr. 169
---

States                                                                 --->      United States
United States v. Rose, 24 CMR 251
---

States                                                                 --->      United States
United States v. Gunnels, 23 CMR 354
---

Angelet                                                                --->      United States ex rel. Angelet
United States ex rel. Angelet v. Fay, 333 F. 2d 12
---

Inc.                                                                   --->      Ziffrin, Inc.
Ziffrin, Inc. v. United States, 318 U. S. 73
---

Morse, 60 Cal. 2d 631                                                  --->      Morse
People v. Morse, 60 Cal. 2d 631, 388 P. 2d 33
---

Morse, 60 Cal. 2d 631, 388 P. 2d 33                                    --->      Morse
People v. Morse, 60 Cal. 2d 631, 388 P. 2d 33, 36 Cal. Rptr. 201
---

States                                                                 --->      United States
United States v. Carignan, 342 U. S. 36
---

Co.                                                                    --->      Shotwell Mfg. Co.
Shotwell Mfg. Co. v. United States, 371 U. S. 341
---

States                                                                 --->      United States
United States v. Robinson, 354 F. 2d 109
---

Townsend                                                               --->      United States ex rel. Townsend
United States ex rel. Townsend v. Ogilvie, 334 F. 2d 837
---

Hartgraves, 31 Ill. 2d 375                                             --->      Hartgraves
People v. Hartgraves, 31 Ill. 2d 375, 202 N. E. 2d 33
---

Fox, ___ Iowa ___                                                      --->      Fox
State v. Fox, ___ Iowa ___, 131 N. W. 2d 684
---

Warden, 236 Md. 236                                                    --->      Warden
Parker v. Warden, 236 Md. 236, 203 A. 2d 418
---

State, ___ Nev. ___                                                    --->      State
Bean v. State, ___ Nev. ___, 398 P. 2d 251
---

Hodgson, 44 N. J. 151                                                  --->      Hodgson
State v. Hodgson, 44 N. J. 151, 207 A. 2d 542
---

Gunner, 15 N. Y. 2d 226                                                --->      Gunner
People v. Gunner, 15 N. Y. 2d 226, 205 N. E. 2d 852
---

Linde                                                                  --->      Commonwealth ex rel. Linde
Commonwealth ex rel. Linde v. Maroney, 416 Pa. 331
---

Linde                                                                  --->      Commonwealth ex rel. Linde
Maroney, 416 Pa. 331                                                   --->      Maroney
Commonwealth ex rel. Linde v. Maroney, 416 Pa. 331, 206 A. 2d 288
---

State, 24 Wis. 2d 491                                                  --->      State
Browne v. State, 24 Wis. 2d 491, 131 N. W. 2d 169
---

Russo                                                                  --->      United States ex rel. Russo
United States ex rel. Russo v. New Jersey, 351 F. 2d 429
---

Dorado, 62 Cal. 2d 338                                                 --->      Dorado
People v. Dorado, 62 Cal. 2d 338, 398 P. 2d 361
---

Dufour, ___ R. I. ___                                                  --->      Dufour
State v. Dufour, ___ R. I. ___, 206 A. 2d 82
---

Neely, 239 Ore. 487                                                    --->      Neely
State v. Neely, 239 Ore. 487, 395 P. 2d 557
---

Neely, 239 Ore. 487, 395 P. 2d 557, modified                           --->      Neely
State v. Neely, 239 Ore. 487, 395 P. 2d 557, modified, 398 P. 2d 482
---

Dorado, 62 Cal. 2d 338                                                 --->      Dorado
People v. Dorado, 62 Cal. 2d 338, 398 P. 2d 361
---

States                                                                 --->      United States
United States v. Bolden, 355 F. 2d 453
---

Du Bont, 235 Cal. App. 2d 844                                          --->      Du Bont
```

I ran a test on Miranda v. Arizona to illustrate the improvements.  It lists only changes identified.  

In this case you can see accurate plaintiff parsing and you can also see better defendant parsing.  Especially when or usually when the citation is a parallel to the main.  It no longer mistakenly identifies the previous citation with the defendant name when handling markup text.

grossir reviewed (commented):
The names that this PR are able to collect look great. 

However, there are some weird losses on the benchmark

- [Akins](https://www.courtlistener.com/opinion/4989605/doe-v-poritz/)

Since it's picking everything inside the emphasis tag, it is picking 'Cf' as part of the name. This makes it lose the following `Akins` standalone reference, since it's looking for `Cf. Akins`. Solvable by cleaning up stopwords like `Cf`. Another one that happens a lot inside the same style tags as the name is "See", you should probably include these in the tests

```
 FullCaseCitation('104 So.3d 1173', groups={'volume': '104', 'reporter': 'So.3d', 'page': '1173'}, metadata=FullCaseCitation.Metadata(parenthetical='reversing nonhomicide sentence of life without parole where trial court exercised its discretion in imposing a sentence of twenty-seven years on the related homicide offense of second-degree murder, an offense punishable by a maximum of life in prison', pin_cite=None, pin_cite_span_start=None, pin_cite_span_end=None, year='2012', court=None, plaintiff='Cf. Akins', defendant='State', extra=None, antecedent_guess=None, resolved_case_name_short=None, resolved_case_name=None)),
```


- Foy

This is the full citation "The opinion in Thrift Funds Canal, Inc. v. Foy, [261 La. 573](https://www.courtlistener.com/opinion/1890117/thrift-funds-canal-inc-v-foy/), [260 So.2d 628](https://www.courtlistener.com/opinion/1890117/thrift-funds-canal-inc-v-foy/) (1972), classifies conventional mortgages into three categories:"

On this branch, it's not picking up "Foy" as defendant
```
 FullCaseCitation('261 La. 573', groups={'volume': '261', 'reporter': 'La.', 'page': '573'}, metadata=FullCaseCitation.Metadata(parenthetical=None, pin_cite=None, pin_cite_span_start=None, pin_cite_span_end=None, year='1972', court=None, plaintiff='Inc.', defendant=None, extra='260 So.2d 628', antecedent_guess=None, resolved_case_name_short=None, resolved_case_name=None)),
 FullCaseCitation('260 So.2d 628', groups={'volume': '260', 'reporter': 'So.2d', 'page': '628'}, metadata=FullCaseCitation.Metadata(parenthetical=None, pin_cite=None, pin_cite_span_start=None, pin_cite_span_end=None, year='1972', court=None, plaintiff='Inc.', defendant=None, extra=None, antecedent_guess=None, resolved_case_name_short=None, resolved_case_name=None)),
```

but on main, it is picking it up. I am not sure about the reason for the change...

```
 FullCaseCitation('261 La. 573', groups={'volume': '261', 'reporter': 'La.', 'page': '573'}, metadata=FullCaseCitation.Metadata(parenthetical=None, pin_cite=None, pin_cite_span_start=None, pin_cite_span_end=None, year='1972', court=None, plaintiff='Inc.', defendant='Foy', extra='260 So.2d 628', antecedent_guess=None, resolved_case_name_short=None, resolved_case_name=None)),
 FullCaseCitation('260 So.2d 628', groups={'volume': '260', 'reporter': 'So.2d', 'page': '628'}, metadata=FullCaseCitation.Metadata(parenthetical=None, pin_cite=None, pin_cite_span_start=None, pin_cite_span_end=None, year='1972', court=None, plaintiff='Inc.', defendant='Foy', extra=None, antecedent_guess=None, resolved_case_name_short=None, resolved_case_name=None)),
```

Which makes it able to pick up 2 references below

![image](https://github.com/user-attachments/assets/2b60ebec-bdb1-455f-a948-78d62ca666f1)


You should probably check the "gain" column (which is actually the loss)

flooie commented:
1. ` Judge Regan (dissenting) in <i>Thrift Funds Canal,</i> Inc. v. Foy, 242 So.2d 253, 257 (La.App. 4 Cir. 1970), calls` 

we lose FOY because its not correctly tagged in HTML

flooie commented on eyecite/helpers.py:
```diff
@@ -129,28 +132,49 @@ def add_post_citation(citation: CaseCitation, words: Tokens) -> None:
                 ).strip("( ")
                 offset += len(citation.metadata.plaintiff) + 1
+
+                if (
+                    document.markup_text
+                    and document.plain_to_markup
+                    and document.markup_to_plain
+                ):
```
this was done to satisfy mypy

flooie commented on eyecite/helpers.py:
```diff
@@ -440,6 +475,85 @@ def filter_citations(citations: List[CitationBase]) -> List[CitationBase]:
+    """
+    Use markup to identify if the plaintiff/antecedent guess is incomplete.
+
+    Parameters:
+        document: The Document instance.
+        base_offset: The plain text offset to convert
+        default_text: The current text stored for this field.
+        source_end: Optional plain text end index to use when extracting text.
+                    If provided, full text is extracted as:
+                        document.plain_text[plain_text_start:source_end]
```
actually I think the format is somewhere in the middle 

```    """This is eyecite's main workhorse function. Given a string of text
    (e.g., a judicial opinion or other legal doc), return a list of
    `eyecite.models.CitationBase` objects representing the citations found
    in the doc.

    Args:
        plain_text: The text to parse. You may wish to use the
            `eyecite.clean.clean_text` function to pre-process your text
            before passing it here.
        remove_ambiguous: Whether to remove citations that might refer to more
            than one reporter and can't be narrowed down by date.
        tokenizer: An instance of a Tokenizer object. See `eyecite.tokenizers`
            for information about available tokenizers. Uses the
            `eyecite.tokenizers.AhocorasickTokenizer` by default.
        markup_text: if the source text has markup (XML or HTML mostly), pass
            it to extract ReferenceCitations that may be detectable via
            markup style tags
        clean_steps: Cleanup steps and methods

    Returns:
        A list of `eyecite.models.CitationBase` objects
```

flooie commented on eyecite/helpers.py:
```diff
@@ -174,9 +200,18 @@ def add_defendant(citation: CaseCitation, words: Tokens) -> None:
+                and document.markup_to_plain
+            ):
+
+                new_text = update_defendant_markup(document, stop_word)
+                if citation.metadata.defendant != new_text:
+                    citation.metadata.defendant = new_text
```
this wont affect the start, this actually just reduces the defendant on the back end.  Thats why no update.

flooie commented on eyecite/helpers.py:
```diff
@@ -440,6 +475,85 @@ def filter_citations(citations: List[CitationBase]) -> List[CitationBase]:
+    return plain_text_start, cleaned_text
+
+
+def update_defendant_markup(document, stop_word):
```
added

flooie commented on eyecite/helpers.py:
```diff
@@ -440,6 +475,85 @@ def filter_citations(citations: List[CitationBase]) -> List[CitationBase]:
+def update_defendant_markup(document, stop_word):
+    # if we have markup - we may want to trim the end of defendant
+    markup_start = document.plain_to_markup.update(
+        stop_word.start + 3, bisect_right
```
Added a comment

flooie commented on eyecite/helpers.py:
```diff
@@ -129,28 +132,49 @@ def add_post_citation(citation: CaseCitation, words: Tokens) -> None:
+                    updated_start, cleaned_text = (
+                        extract_full_text_from_markup(
+                            document,
+                            stop_word.start - 2,
```
added comment - but to ensure we are in the next tag or word

grossir reviewed (commented):
This looks great, I think it's mostly ready

My only nitpick I is that `add_defendant` is now a big and complex function, and adds both plaintiff and defendant, and we should refactor it eventually

mlissner commented:
> My only nitpick is that add_defendant is now a big and complex function, and adds both plaintiff and defendant, and we should refactor it eventually

I haven't looked at the PR, but I think now may be the best time to clean things up, since we probably won't return to eyecite for a while, and when we do it'll be harder to do the refactoring.

flooie commented:
> > My only nitpick is that add_defendant is now a big and complex function, and adds both plaintiff and defendant, and we should refactor it eventually
> 
> I haven't looked at the PR, but I think now may be the best time to clean things up, since we probably won't return to eyecite for a while, and when we do it'll be harder to do the refactoring.

Just seeing this.  It did both plaintiff and defendant previously.  But plaintiff was always a throw-a-way process.

flooie commented:
@grossir something is cached or broken about the benchmark ... I found one in particular weird but others I couldn't replicate.  

this entry ```<!DOCTYPE html>
1744543 |   | I am not going to let you represent yourself
-- | -- | --


```

made no sense - so I ran it locally to test it out 

[
    {
        "cites": [
            "\u00a7",
            "\u00a7",
            "\u00a7",
            "422 U.S. 806",
            "95 S.Ct. 2525",
            "45 L.Ed.2d 562",
            "\u00a7",
            "542 S.W.2d 549",
            "685 S.W.2d 942",
            "587 S.W.2d 611",
            "589 F.2d 45",
            "77 Cal.App.3d 45",
            "143 Cal.Rptr. 379",
            "270 Ind. 55",
            "383 N.E.2d 309",
            "267 Ind. 289",
            "370 N.E.2d 318",
            "95 Nev. 933",
            "604 P.2d 338",
            "590 P.2d 679",
            "588 P.2d 572",
            "98 A.L.R.3d 13",
            "\u00a7",
            "572 S.W.2d 440",
            "622 S.W.2d 416",
            "\u00a7",
            "\u00a7",
            "\u00a7",
            "\u00a7",
            "647 S.W.2d 799",
            "\u00a7",
            "558 S.W.2d 646",
            "593 S.W.2d 208",
            "\u00a7",
            "694 S.W.2d 903",
            "603 S.W.2d 569",
            "676 S.W.2d 304",
            "664 S.W.2d 617",
            "463 U.S. 277",
            "103 S.Ct. 3001",
            "77 L.Ed.2d 637",
            "supra,",
            "\u00a7\u00a7",
            "\u00a7\u00a7",
            "\u00a7",
            "\u00a7",
            "\u00a7\u00a7",
            "\u00a7\u00a7",
            "\u00a7\u00a7",
            "\u00a7\u00a7",
            "\u00a7\u00a7",
            "\u00a7\u00a7",
            "\u00a7\u00a7",
            "\u00a7",
            "\u00a7",
            "\u00a7\u00a7",
            "\u00a7",
            "\u00a7",
            "\u00a7",
            "459 U.S. 359",
            "103 S.Ct. 673",
            "74 L.Ed.2d 535",
            "647 S.W.2d 799",
            "624 S.W.2d 16",
            "689 S.W.2d 758",
            "657 S.W.2d 677",
            "653 S.W.2d 227",
            "supra,",
            "664 S.W.2d 617",
            "603 S.W.2d 683",
            "588 S.W.2d 65"
        ],
        "time": 0.243268
    }
]

and this was the output - you can see there is no random long string.

flooie commented:
@grossir there we go -  I need to re run name extraction because many html files have some tagged nicely and some not.

flooie commented:
| **Old Data**                                                                       | **New Extraction**                                                          |
|--------------------------------------------------------------------------------------|-----------------------------------------------------------------------------|
| Inc. v. Toups,                                                                       | E.g., New Orleans Silversmiths, Inc. v. Toups,                               |
| Inc. v. Foy,                                                                         | Thrift Funds Canal, Inc. v. Foy,                                             |
| Inc. v. Foy, 261 La. 573,                                                              | Thrift Funds Canal, Inc. v. Foy, 261 La. 573,                                |
| cit.                                                                                 | cit.                                                                        |
| Foy,                                                                                 | Foy,                                                                        |
| cit.                                                                                 | cit.                                                                        |
| Toups,                                                                               | Toups,                                                                      |
| Odom v. Cherokee Homes, Inc.,                                                        | Odom v. Cherokee Homes, Inc.,                                               |
| Inc. v. I.D.S. Mortgage Corp.,                                                        | Slidell Building Supply, Inc. v. I.D.S. Mortgage Corp.,                      |
| Inc. v. I.D.S. Mortgage Corp., 273 So.2d 343 (La.App. 1 Cir. 1972), writ den. La.,     | La.,                                                                        |
| Company v. Subco, Inc.,                                                              | of Baton Rouge Bank and Trust Company v. Subco, Inc.,                        |
| Foy,                                                                                 | Foy,                                                                        |
| cit.                                                                                 | cit.                                                                        |
| Union v. Montgomery,                                                                 | League Central Credit Union v. Montgomery,                                  |
| Union v. Montgomery, 251 La. 971,                                                      | League Central Credit Union v. Montgomery, 251 La. 971,                      |
| Monroe v. Carson Homes, Inc.,                                                        | Monroe v. Carson Homes, Inc.,                                               |
| Reed v. Meaux,                                                                       | Reed v. Meaux,                                                              |
| Corporation v. Breaux,                                                               | Chrysler Credit Corporation v. Breaux,                                      |
| Tapp v. Guaranty Finance Co.,                                                        | Tapp v. Guaranty Finance Co.,                                               |
| Union v. Montgomery, supra, at                                                         | Montgomery,                                                                 |
| Co. v. Cappel and Curry,                                                             | Miller, Lyon and Co. v. Cappel and Curry,                                   |
| Rocchi v. Keen,                                                                      | Li Rocchi v. Keen,                                                          |
| Rocchi v. Keen, 242 La. 111,                                                           | Li Rocchi v. Keen, 242 La. 111,                                              |
| Buckner v. Carmack,                                                                  | Buckner v. Carmack,                                                         |
| Company v. Mones,                                                                    | Myrtle Grove Packing Company v. Mones,                                     |
| Company v. Mones, 226 La. 287,                                                         | Myrtle Grove Packing Company v. Mones, 226 La. 287,                         |
| Co. v. Cappel and Curry,                                                             | Co. v. Cappel and Curry,                                                    |
| Rouge v. Heroman,                                                                    | Louisiana National Bank of Baton Rouge v. Heroman,                          |
| Scarborough v. Duke,                                                                 | Scarborough v. Duke,                                                        |
| Foy,                                                                                 | Foy,                                                                        |
| Inc.,                                                                               | Inc.,                                                                      |
| Company v. Soileau,                                                                  | E.g., Ford Motor Credit Company v. Soileau,                                  |
| Breaux,                                                                              | Breaux,                                                                   |
| Inc. v. Vicaro,                                                                      | Polk Chevrolet, Inc. v. Vicaro,                                             |
| Breaux,                                                                              | Breaux,                                                                   |
| Inc. v. Foy,                                                                         | Thrift Funds Canal, Inc. v. Foy,                                            |
| the Court of Appeal (for different reasons),                                         |                                                                             |
| the Court of Appeal (for different reasons), 261 La. 573,                            |                                                                             |

flooie commented:
I added a list of old vs new name extraction for a random case.  using only plain text.

grossir commented on eyecite/helpers.py:
```diff
@@ -129,54 +131,378 @@ def add_post_citation(citation: CaseCitation, words: Tokens) -> None:
+        bisect_right,
+    )
+    case_name = document.plain_text[start:end]
+    # print("------->", case_name, start, markup_location)
```
delete commented print

grossir commented on tests/test_AnnotateTest.py:
```diff
@@ -313,8 +313,8 @@ def test_span_with_pincite(self):
             (
-                "principal residence.” Nobelman at 332, 113 S.Ct. 2106",
-                ["Nobelman at 332, 113 S.Ct. 2106"],
+                "principal residence.” Nobelman 1 U.S. 1 at 332, 113 S.Ct. 2106",
```
I think the change in this test actually breaks the intention of the test. Why change it?

flooie replied:
we spoke, this test was broken ... Ive removed it completely

grossir commented on tests/test_FindTest.py:
```diff
@@ -458,15 +458,41 @@ def test_find_citations(self):
+            ('(Mo.); Bean v. State, — Nev. —, 398 P. 2d 251; ',
+             [case_citation(volume='398', reporter='P. 2d', page='251',
+                            metadata={'plaintiff': 'Bean',
+                                      'defendant': 'State, — Nev. —'})]
```
defendant should just be 'State'?

flooie replied:
yes - I just didnt solve this case - so I left it for now.

flooie replied:
sounds good ... I have a solution

grossir commented on tests/test_FindTest.py:
```diff
@@ -458,15 +458,41 @@ def test_find_citations(self):
               ]),
             # Square brackets around year
-            ('Rogers v Rogers (63 NY2d 582 [1984])',
+            ('Rogers v Rogers 63 NY2d 582 [1984]',
```
I think changing this test is actually breaking a live data example?
It was recently introduced
https://github.com/freelawproject/eyecite/commit/2975617cfc634718b0c6707bee412888e3b22f52#diff-dfee4fd2e40b9d83c093548058ae0b3b0b5f5c378b4fc3819ad02c199afb248d

flooie replied:
added back in

grossir commented on tests/test_FindTest.py:
```diff
@@ -458,15 +458,41 @@ def test_find_citations(self):
+                            metadata={'plaintiff': 'Spano',
+                                      'defendant': 'People of State of New York'})]
+             ),
+            # Spano v. People of State of New York, 360 U.S. 315, 321, n. 2, 79 S.Ct. 1202, 1206, 3 L.Ed.2d 1265, collects 28 cases.
```
repeated tests

grossir commented on tests/test_FindTest.py:
```diff
@@ -779,8 +793,8 @@ def test_find_tc_citations(self):
                             volume='2018')]),
-            ('U.S. 1234 1 U.S. 1',
-             [case_citation(volume='1', reporter='U.S.', page='1')]),
+            # ('U.S. 1234 1 U.S. 1',
```
delete commented line

flooie replied:
ok

grossir commented on tests/test_FindTest.py:
```diff
@@ -1000,7 +1014,7 @@ def test_citation_fullspan(self):
             "42 U.S.C. § 405(r)(2) (2019)",
             "37 A.L.R.4th 972, 974 (1985)",
             "497 Fed. Appx. 274 (4th Cir. 2012)",
-            "Corp. v. Nature's Farm Prods., No. 99 Civ. 9404 (SHS), 2000 U.S. Dist. LEXIS 12335 (S.D.N.Y. Aug. 25, 2000)",
```
why is this being deleted / commented?

flooie replied:
it was deleted and put back.  it was a tough case to solve but should be fixed now

grossir commented on eyecite/helpers.py:
```diff
@@ -129,54 +131,378 @@ def add_post_citation(citation: CaseCitation, words: Tokens) -> None:
+            title_starting_index = index - 1
+            pre_cite_year = str(word)[1:5]
+            continue
+        if (
```
there are a lot of `str(word)` calls in the if clauses, may as well do it once and use a variable

grossir reviewed (commented):
From the changed tests, I think the changes may cause us to miss a few citations we were already able to process

For the losses (Gain column), the "Sime" losses are sort of OK, they were a full case name reference like "Sime v Malouf", where both Sime and Malouf where identified as references. Now Sime is lost, so it's deduplicating, but still not quite right

"[Reece](https://www.courtlistener.com/opinion/280035/elmo-williams-v-h-v-field/)" is an actual loss
The current branch fails to extract any party names, cascading into a missing reference
```
 FullCaseCitation('310 F.2d 139', groups={'volume': '310', 'reporter': 'F.2d', 'page': '139'}, metadata=FullCaseCitation.Metadata(parenthetical=None, pin_cite=None, pin_cite_span_start=None, pin_cite_span_end=None, year='1962', court='ca9', plaintiff=None, defendant=None, extra=None, antecedent_guess='Washington', resolved_case_name_short=None, resolved_case_name=None)),
```

Main does OK
```
 FullCaseCitation('310 F.2d 139', groups={'volume': '310', 'reporter': 'F.2d', 'page': '139'}, metadata=FullCaseCitation.Metadata(parenthetical=None, pin_cite=None, pin_cite_span_start=None, pin_cite_span_end=None, year='1962', court='ca9', plaintiff='Reece', defendant='State of Washington', extra=None, antecedent_guess=None, resolved_case_name_short=None, resolved_case_name=None)),
```

-------------

Not directly related to these changes:

I found an error that may be bug with `Document.clean_text`. Cleaning `xml_harvard` will break the extraction because it's missing the xml tags cleanup step.  This has been done outside the clean steps, but we may as well handle it too inside `Document.clean_text`, to make the usage easier? To reproduce:
```python
import requests
from eyecite import get_citations


opid = 280035
jason = requests.get("https://www.courtlistener.com/api/rest/v4/opinions/{}/".format(opid)).json()
citations = get_citations(markup_text=jason['xml_harvard'], clean_steps=["html", "all_whitespace"])
```

flooie commented:
6596585 | Beckwith |   - is a good loss 
2414924 | Boerne at 2170 - is a good loss 

The only loses are the two improvements

grossir reviewed (commented):
I took another sample, I think we may be overfitting the static sample that we have. I am attaching a partial output file that shows a lot of gains, but some unexplained loses (in this case, the columns are named correctly)

[new_sample_output.csv](https://github.com/user-attachments/files/19786071/new_sample_output.csv)

In general, I think the changes are helping catch names, but we may as well keep the ones we were already able to catch.

Apart from that, is it possible to refactor the `for` loops for the 2 main changes `find_case_name` (200 lines) and `find_case_name_in_html` (121 lines)? They are both too long, which makes them hard to read / understand

flooie commented:
From what I can tell all of the losses now are actually correct.

flooie commented on eyecite/models.py:
```diff
@@ -521,21 +521,29 @@ class Metadata(CaseCitation.Metadata):
+            find_case_name(self, document)
+
+        if self.metadata.defendant is None:
+            find_case_name(self, document)
```
yep

grossir commented on eyecite/helpers.py:
```diff
@@ -358,144 +512,323 @@ def find_html_tags_at_position(
     citation: CaseCitation, document: Document, short: bool = False
-):
-    """Add case name from HTML
+) -> Optional[None]:
```
Just `-> None` as return type should be enough?

grossir commented on CHANGES.md:
```diff
@@ -9,11 +9,13 @@ Features:
 - Simplifies citation processing by reducing parameter passing and improving maintainability (hopefully).
 - Should enable more complex html parsing.
 - Adds support for years preceding citations
+- Improve markup plaintiff extraction
```
Should a line about "placeholder" regex and token for capturing placeholder citations

grossir reviewed (commented):
@flooie I think this is ready for merging

The refactor does help readibility, and you made some nice additions 
- extra stop words
- start pagination bugs

Just some minor comments below

mlissner commented:
Congrats all! I can't wait to see this in action.
